src/Elastic.Apm/Report/PayloadSenderV2.cs

```diff
@@ -55,6 +55,7 @@ public PayloadSenderV2(IApmLogger logger, IConfigurationReader configurationRead
 			servicePoint.ConnectionLimit = 20;
 
 			_httpClient = new HttpClient(handler ?? new HttpClientHandler()) { BaseAddress = serverUrlBase };
+			_httpClient.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue(new ProductHeaderValue("elasticapm-dotnet", _service.Agent.Version)));
```

Comment:
Shouldn't we add `http-lib/version` and `language/version` parts while we are at it? How much more effort would you estimate it is?
I think for 
- `http-lib/version` we can use `System.Net.Http.HttpClient`/<System.Net.Http.HttpClient assembly version?>
- `language/version` we can use [`Elastic.Apm.Api.Runtime.Name`](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Api/Service.cs#L81)/[`Elastic.Apm.Api.Runtime.Version`](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Api/Service.cs#L84)

What do you think?

Reply:
I didn't add `http-lib/version` and `language/version` due to there aren't implemented in all agents. However, it isn't a problem. Will do it,

Reply:
@SergeyKleyman, I need a quick consult with you about `language/version`. `ProductInfoHeaderValue` has a validation of each character. This validation can be found at [`HttpRuleParser`](https://github.com/dotnet/corefx/blob/e64cac6dcacf996f98f0b3f75fb7ad0c12f588f7/src/System.Net.Http/src/System/Net/Http/HttpRuleParser.cs#L63).

`Elastic.Apm.Api.Runtime.Name` can be `.NET Core`, `.NET Framework` or `N/A` which contains illegal symbols (space and `/`). Resolution of this can be done by 3 ways:

1. analyze each character
```csharp
string AdaptProductName(string productName)
{
    var x = new HashSet<char> { ' ', '(', ')', '<', '>', '@', ',', ';', ':', '\\', '"', '/', '[', ']', '?', '=', '{', '}', (char)127 /*DEL*/ };

    var chars = productName.ToCharArray();

    for (var i = 0; i < chars.Length; i++)
    {
        if (x.Contains(chars[i]))
	    chars[i] = '-';
    }
    return new string(chars);
}
```
 
2. use `Regex.Replace`
3. use `Elastic.Apm.Api.Runtime.Name.Replace(' ', '-').Replace('/', '-')` 

What is the preferable way?

Reply:
I would just set `User-Agent` header to a raw string instead of using `ProductInfoHeaderValue`, for example as suggested [here](https://stackoverflow.com/questions/43215527/c-sharp-httpclient-adding-user-agent-header-shows-up-as-several-different-head).

Reply:
I thought a little bit more about `http-lib/version` part (for which we can use `System.Net.Http.HttpClient`/<System.Net.Http.HttpClient assembly version?>) and it seems to me it will be redundant for .NET agent since we use HTTP client provided by the runtime and not a 3rd party library so if we know .NET runtime name/version (which will be in `language`/`version`) that will be enough. What do you think?

Reply:
> I would just set `User-Agent` header to a raw string instead of using `ProductInfoHeaderValue`, for example as suggested [here](https://stackoverflow.com/questions/43215527/c-sharp-httpclient-adding-user-agent-header-shows-up-as-several-different-head).

It can be the case, however, we still need to check all strings for spaces and `/`, because [`space` symbol is used as separator between sub-products](https://tools.ietf.org/html/rfc2616#section-3.8) and `/` symbol is used to separate product name with product version. I don't find mention of other symbols which are included in `HttpRuleParser`, but I'm afraid it's reasonable. Will investiagate it deeper

Reply:
> I thought a little bit more about `http-lib/version` part (for which we can use `System.Net.Http.HttpClient`/<System.Net.Http.HttpClient assembly version?>) and it seems to me it will be redundant for .NET agent since we use HTTP client provided by the runtime and not a 3rd party library so if we know .NET runtime name/version (which will be in `language`/`version`) that will be enough. What do you think?

As I understand, `System.Net.Http` can be references as a [NuGet package](https://www.nuget.org/packages/System.Net.Http/) in application host and by this reason `.NET runtime name/version` cannot guaranteed correct map.

Reply:
I see. In that case let’s include it.

Reply:
> > I would just set `User-Agent` header to a raw string instead of using `ProductInfoHeaderValue`, for example as suggested [here](https://stackoverflow.com/questions/43215527/c-sharp-httpclient-adding-user-agent-header-shows-up-as-several-different-head).
> 
> It can be the case, however, we still need to check all strings for spaces and `/`, because [`space` symbol is used as separator between sub-products](https://tools.ietf.org/html/rfc2616#section-3.8) and `/` symbol is used to separate product name with product version. I don't find mention of other symbols which are included in `HttpRuleParser`, but I'm afraid it's reasonable. Will investiagate it deeper

I see your point - there’re advantages to having this header easily machine parseable. I would go with your suggestion to replace invalid characters. I would replace spaces with underscores and slashes with dashes.